src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -393,6 +393,63 @@ bool RegionNode::is_unreachable_from_root(const PhaseGVN* phase) const {
   return true; // The Region node is unreachable - it is dead.
 }
 
+bool RegionNode::is_self_loop(Node* n) {
+  Node_List nstack;
+  VectorSet visited;
+  nstack.push(n);
+  visited.set(n->_idx);
+
+  while(nstack.size()) {
+    n = nstack.pop();
+    for(unsigned i = 1; i < n->req() ; i++) {
+      Node* in = n->in(i);
+      if (in == NULL) {
+        continue;
+      }
+      if (visited.test_set(in->_idx)) {
+        return true;
+      } else {
+        nstack.push(in);
+        visited.set(in->_idx);
+      }
+    }
+  }
+
+  return false;
+}
+
+// If a two input non-loop region has dead input
+// edge[s] degenerate any phi node contained within it.
+bool RegionNode::try_phi_disintegration(PhaseGVN *phase) {
```

Comment:
RegionNode::try_phi_disintegration - is it a requirement for this enhancement? or a separate issue?

Also - I know we already remove phis that only have one input. If the input is set to top -  PhiNode::Ideal should reduce the phi. If you have found a case where this doesn't happen - we should investigate and fix.

Reply:
This transformation is being done during RegionNode idealization,  A phi may be intact (have both valid inputs), but if its parent region has one control edge connected to top() in that case the phi-input corresponding to top() edge is being removed and phi is disintegrated.  Currently for dead loops all its phi nodes are replaced by top() 

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/cfgnode.cpp#L575

For partial in-lining I introduced a mem phi node at the convergence of fast and slow path during node expansion which was getting replaced by top() during RegionNode idealization in case of a dead loop. This mem_phi had a user outside loop.
  while ( )  {     // dead loop detection 
     if ( len < 32 ) 
         fast_path
     else
         slow_path
     mem_Phi = Memory(fast_path, slow_path)
  }
  memory = MemMerge(mem_Phi);